File: 0902-numbers-at-most-n-given-digit-set/0902-numbers-at-most-n-given-digit-set.py
class Solution:
    def atMostNGivenDigitSet(self, digits: List[str], n: int) -> int:
        N = str(n)
        n_len, d_len = len(N), len(digits)
        res = sum(d_len ** i for i in range(1, n_len))
        i = 0
        while i < n_len:
            res += sum(digit < N[i] for digit in digits) * (d_len ** (n_len - i - 1))
            if N[i] not in digits:
                break
            i += 1
        return res + (i == n_len)
          

File: 0902-numbers-at-most-n-given-digit-set/0902-numbers-at-most-n-given-digit-set.js
/**
 * @param {string[]} digits
 * @param {number} n
 * @return {number}
 */
var atMostNGivenDigitSet = function(digits, n) {
  const str = n.toString();
  const [nLen, dLen] = [str.length, digits.length];
  let [res, i] = [0, 0];
  
  for (let j = 1; j < nLen; j++) res += dLen ** j;
  
  while (i < nLen) {
    res += digits.reduce((acc, curr) => acc + (+curr < +str[i]), 0) * (dLen ** (nLen - i - 1));
    if (!digits.includes(str[i])) break
    
    i++;
  }
  
  return res + (i === nLen);
};